drivers/clock_control/clock_control_r8a7795_cpg_mssr.c

```diff
@@ -20,140 +20,285 @@
+
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD1H, 0x0078, RCAR_CPG_NONE, RCAR_CPG_MHZ(800)),
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD1, 0x0078, R8A7795_CLK_SD1H, RCAR_CPG_MHZ(800)),
+
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD2H, 0x0268, RCAR_CPG_NONE, RCAR_CPG_MHZ(800)),
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD2, 0x0268, R8A7795_CLK_SD2H, RCAR_CPG_MHZ(800)),
+
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD3H, 0x026C, RCAR_CPG_NONE, RCAR_CPG_MHZ(800)),
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_SD3, 0x026C, R8A7795_CLK_SD3H, RCAR_CPG_MHZ(800)),
+
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_CANFD, 0x0244, RCAR_CPG_NONE, RCAR_CPG_MHZ(800)),
+
+	RCAR_CORE_CLK_INFO_ITEM(R8A7795_CLK_S0D12, RCAR_CPG_NONE,
+				RCAR_CPG_NONE, RCAR_CPG_KHZ(66600)),
+};
+
+/* NOTE: the array MUST be sorted by module field */
+static struct cpg_clk_info_table mod_props[] = {
+	RCAR_MOD_CLK_INFO_ITEM(310, R8A7795_CLK_S3D4),
+
+	RCAR_MOD_CLK_INFO_ITEM(311, R8A7795_CLK_SD3),
+	RCAR_MOD_CLK_INFO_ITEM(312, R8A7795_CLK_SD2),
+	RCAR_MOD_CLK_INFO_ITEM(313, R8A7795_CLK_SD1),
+	RCAR_MOD_CLK_INFO_ITEM(314, R8A7795_CLK_SD0),
+};
+
+static int r8a7795_cpg_enable_disable_core(const struct r8a7795_cpg_mssr_config *cfg,
+					struct cpg_clk_info_table *clk_info,
+					uint32_t enable)
 {
-	uint32_t divider;
-	unsigned int key;
 	int ret = 0;
+	uint32_t reg;
 
-	/* Only support CANFD core clock at the moment */
-	if (module != R8A7795_CLK_CANFD) {
-		return -EINVAL;
+	enable = !!enable;
+
+	switch (clk_info->module) {
+	case R8A7795_CLK_SD0:
+	case R8A7795_CLK_SD1:
+	case R8A7795_CLK_SD2:
+	case R8A7795_CLK_SD3:
+		reg = sys_read32(cfg->base_address + clk_info->offset);
+		reg &= ~R8A7795_CLK_SD_STOP_BIT;
+		reg |= !enable * R8A7795_CLK_SD_STOP_BIT;
+		break;
+	case R8A7795_CLK_SD0H:
+	case R8A7795_CLK_SD1H:
+	case R8A7795_CLK_SD2H:
+	case R8A7795_CLK_SD3H:
+		reg = sys_read32(cfg->base_address + clk_info->offset);
+		reg &= ~R8A7795_CLK_SDH_STOP_BIT;
+		reg |= !enable * R8A7795_CLK_SDH_STOP_BIT;
+		break;
+	case R8A7795_CLK_CANFD:
+		reg = sys_read32(cfg->base_address + clk_info->offset);
+		reg &= ~R8A7795_CLK_CANFD_STOP_BIT;
+		reg |= !enable * R8A7795_CLK_CANFD_STOP_BIT;
+		break;
+	default:
+		ret = -ENOTSUP;
+		break;
+	}
+
+	if (!ret) {
+		rcar_cpg_write(cfg->base_address, clk_info->offset, reg);
 	}
+	return ret;
+}
+
+static int r8a7795_cpg_core_clock_endisable(const struct device *dev,
+					    struct rcar_cpg_clk *clk,
+					    bool enable)
+{
+	struct cpg_clk_info_table *clk_info;
+	const struct r8a7795_cpg_mssr_config *cfg = dev->config;
+	struct r8a7795_cpg_mssr_data *data = dev->data;
+	k_spinlock_key_t key;
+	int ret = 0;
 
-	key = irq_lock();
+	clk_info = rcar_cpg_find_clk_info_by_module_id(dev, clk->domain, clk->module);
+	if (!clk_info) {
+		return -EINVAL;
+	}
 
 	if (enable) {
-		if (rate > 0) {
-			if ((CANFDCKCR_PARENT_CLK_RATE % rate) != 0) {
-				LOG_ERR("Can not generate %u from CANFD parent clock", rate);
-				ret = -EINVAL;
-				goto unlock;
-			}
+		if ((int)clk->rate > 0) {
```

Comment:
Clock rates slightly above 2GHz will be happy :)

Reply:
You already has 3GHz for S4. And S4 is not the last SoC in Gen4 series. It is the first, actually. I  believe there will be SoCs with  much more rich peripherals and more complex clock tree. So, there are high chances that you will need to rework this code in the very near future.

I am strongly suggesting to make it right from the beginning.